Add receive optout hook endpoint and task

rudigiesler opened the pull request:
This PR adds a secured API endpoint which can receive rapidpro hooks.

This endpoint will accept the hook, and then run a background task for processing the optout.

It also adds an element to the UI for viewing the list of opt outs, and their processing status

![image](https://user-images.githubusercontent.com/8234653/41473714-5f615e80-70ba-11e8-8683-d2934bfc5866.png)

rudigiesler committed:
Add opt out model + views + API for receiving opt out hooks

rudigiesler committed:
Add documentation for opt out API

rudigiesler committed:
Add background task for processing opt outs

OMosimege reviewed (approved):
This looks great Rudi 💯 you can merge, but I'll probably keep asking questions as I learn what each line does one by one

jerith commented:
Looks like you need some more test coverage of the newly added bits. :-)

jerith commented on lib/companion/application.ex:
```diff
@@ -12,9 +15,23 @@ defmodule Companion.Application do
+      {Honeydew.EctoPollQueue,
+       [
+         :process_opt_out,
+         schema: OptOut,
```
Maybe pull this `Honeydew.EctoPollQueue` config out into a separate function (or module attribute or something) and just reference it here? Having a big structure in the middle of a list of one-line child process definitions is a bit disorienting and the `Honeydew.Workers` at the bottom gets a bit lost.

jerith commented:
You've added direct references to `Poison`, but you don't explicitly depend on it -- it's pulled in by other libraries.

It's generally a good idea to only reference direct dependencies, otherwise changes in some other library (several projects seem to be switching from `Poison` to `Jason`, for example) might suddenly leave you without one of your dependencies installed.

(One of the things I really like about Rust is that the compiler won't actually let you do this -- only modules from libraries that you directly depend on are available to your code. I wonder if it's possible for a tool to check this in Elixir...)

rudigiesler committed:
Move supervisor config for task into task

rudigiesler committed:
Add test for docs

rudigiesler committed:
Add test for format_timestamp view function

rudigiesler committed:
Add test for process opt out failure

rudigiesler committed:
Depend on poison directly

rudigiesler commented:
It seems like most of the coveralls deficit is due to the swagger definitions, which I haven't been able to find a way to run during the tests.

rudigiesler commented:
@jerith I think I've addressed all of your comments, and increased the test coverage (all of what's left seems to be the swagger definitions, which I'm not sure how to run in the tests, they only seem to be run when generating the swagger.json). Ready for another look.

rudigiesler committed:
Switch to pipe format for optout task

rudigiesler committed:
Use pipe operator for date formatting in opt out task

jerith commented on lib/companion_web/controllers/fallback_controller.ex:
```diff
@@ -0,0 +1,20 @@
+    conn
+    |> put_status(:unprocessable_entity)
+    |> render(CompanionWeb.ChangesetView, "error.json", changeset: changeset)
+  end
```
Could there be any sensitive data in the changeset that we don't want to return in an error response?

rudigiesler replied:
No I don't think so, everything in the changeset is supplied by the user of the API. The view also extracts the errors from the changeset and just returns that using https://hexdocs.pm/ecto/Ecto.Changeset.html#traverse_errors/2
